lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +324,317 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
+
+Optional<std::vector<int>>
+TaskMonitor::readFromReadyFdsReturningFinishedOnes() {
+  std::vector<int> finishedFds;
+  for (struct pollfd &fd : FdsBeingPolled) {
+    const int fileDes = fd.fd;
+    const short receivedEvents = fd.revents;
+    fd.revents = 0;
+    verifyEvents(receivedEvents);
+    Task &T = TasksBeingExecuted.findTaskForFd(fileDes);
+    readDataIfAvailable(receivedEvents, fileDes, T);
+    if (!didTaskHangup(receivedEvents))
+      continue;
+    finishedFds.push_back(fileDes);
+    const bool hadError = cleanUpAHungUpTask(
+        T, callbacks.TaskFinished, callbacks.TaskSignalled);
+    TasksBeingExecuted.destroyTask(T);
+    if (hadError)
+      return None;
+  }
+  return finishedFds;
+}
+
+void TaskMonitor::verifyEvents(const short events) const {
+  if (events & POLLNVAL) {
+    // We passed an invalid fd; this should never happen,
+    // since we always mark fds as finished after calling
+    // Task::finishExecution() (which closes the Task's fd).
+    llvm_unreachable("Asked poll() to watch a closed fd");
+  }
+  const short expectedEvents = POLLIN | POLLPRI | POLLHUP | POLLERR;
+  assert((events & ~expectedEvents) == 0 && "Received unexpected event");
+}
+
+void TaskMonitor::readDataIfAvailable(const short events, const int fd,
+                                      Task &T) const {
+  if (events & (POLLIN | POLLPRI)) {
+    // There's data available to read. Read _some_ of it here, but not
+    // necessarily _all_, since the pipe is in blocking mode and we might
+    // have other input pending (or soon -- before this subprocess is done
+    // writing) from other subprocesses.
+    //
+    // FIXME: longer term, this should probably either be restructured to
+    // use O_NONBLOCK, or at very least poll the stderr file descriptor as
+    // well; the whole loop here is a bit of a mess.
+    T.readFromPipes(/*UntilEnd*/ false);
   }
+}
+
+bool TaskMonitor::didTaskHangup(const short events) const {
+  return (events & (POLLHUP | POLLERR)) != 0;
+}
+
+static bool cleanUpAHungUpTask(
+    Task &T, const TaskQueue::TaskFinishedCallback FinishedCallback,
+    const TaskQueue::TaskSignalledCallback SignalledCallback) {
+  const Optional<int> StatusIfOK = waitForPid(T.getPid());
+  if (!StatusIfOK)
+    return true;
+
+  T.finishExecution();
+  int Status = *StatusIfOK;
+  return WIFEXITED(Status)
+             ? cleanUpAfterExit(Status, T, FinishedCallback)
+             : WIFSIGNALED(Status)
+                   ? cleanUpAfterSignal(Status, T, SignalledCallback)
+                   : false /* Can this case ever happen? */;
+}
+
+static Optional<int> waitForPid(const pid_t pidToWaitFor) {
+  for (;;) {
+    int Status = 0;
+    const pid_t pidFromWait = waitpid(pidToWaitFor, &Status, 0);
+    if (pidFromWait == pidToWaitFor)
+      return Status;
+    if (pidFromWait > 0)
+      llvm_unreachable(
+          "We asked to wait for this Task, but we got another process ID!");
```

Comment:
`llvm_unreachable` is an optimization hint in release builds, so this is actually *stronger* than the assertion that was here before. If you really want to leave the check in for a release build, you'll have to call `abort` directly…but I don't think you should do that. `assert` is fine.

(this applies here and to the other things you've turned into `llvm_unreachable`)